rootfs/etc/nginx/lua/defer_to_timer.lua

```diff
@@ -0,0 +1,39 @@
+local queue = {}
+local MAX_QUEUE_SIZE = 10000
+local FLUSH_INTERVAL = 1
+
+local _M = {}
+
+local function flush_queue()
+  local current_queue = queue
+  queue = {}
+
+  for _,v in ipairs(current_queue) do
+    v.func(unpack(v.args))
+  end
+end
+
+function _M.init_worker()
+  local _, err = ngx.timer.every(FLUSH_INTERVAL, flush_queue)
```

Comment:
Do you need to care about recursion? If the queues are full (under load) it's conceivable that `flush_queue` could be executing concurrently, working off two separate queues. This could cause events to be generated out of order.

Reply:
@csfrancis I think this can happen with what we currently have as well. Imagine following:
1. A new pending timer gets created at https://github.com/Shopify/nginx-routing-modules/blob/master/lua/shopify/util/defer.lua#L41
2. timer hasn't started yet so queue size will be > 0 and therefore `timer_started` will be set to `true`: https://github.com/Shopify/nginx-routing-modules/blob/master/lua/shopify/util/defer.lua#L46
3. Some items gets added to the queue
4. The pending timer finally starts and sets `timer_started` to `false` and starts working off the queue: https://github.com/Shopify/nginx-routing-modules/blob/master/lua/shopify/util/defer.lua#L18
5. Because `timer_started` is `false` another timer can be created while the current one still processing previous queue
6. The two timer can be running at the same time, one working off old queue and the new one working off items from the new queue, hence events being delivered out of order.

Reply:
I'll add a comment to make it clearer that this module does not guarantee ordering.

Currently there's no consumer of this that'd care about the order.

Reply:
I think you're right (not saying the current behaviour guarantees order either), but I wanted to bring it up as a concern.